src/by/epam/ayem/module5/service/TreasureHouseService.java

```diff
@@ -0,0 +1,125 @@
+    private final static String FILE_NAME = "treasures.txt";
+
+    public void findTreasuresAmountingTo(TreasureHouse treasureHouse) {
+        System.out.println("Please, enter the worth of the treasures: ");
+
+        while (!scanner.hasNextDouble()) {
+            scanner.next();
+            System.out.println("Please, use a number.");
+        }
+
+        double worth = scanner.nextDouble();
+        double totalAmount = 0.0;
+        double left = worth;
+        System.out.println("Treasures amounting to " + worth + ":");
+
+
+        if (!treasureHouse.getTreasures().isEmpty()) {
+            double leastExpensiveTreasure = findLeastExpensiveTreasure(treasureHouse).getTreasureCost();
+
+            while (left >= leastExpensiveTreasure && !treasureHouse.getTreasures().isEmpty()) {
+                int index = random.nextInt(treasureHouse.getTreasures().size());
+
+                if (left >= treasureHouse.getTreasures().get(index).getTreasureCost()) {
+                    totalAmount += treasureHouse.getTreasures().get(index).getTreasureCost();
+                    left = worth - totalAmount;
+                    System.out.println(treasureHouse.getTreasures().get(index));
+                    treasureHouse.getTreasures().remove(index);
+                }
+            }
+            System.out.println("Total worth = " + String.format("%.2f", totalAmount));
+        } else {
+            System.out.println("The treasure house is empty");
+        }
+
+    }
+
+    private Treasure findLeastExpensiveTreasure(TreasureHouse treasureHouse) {
+        double leastExpensive = Double.MAX_VALUE;
+        Treasure leastExpensiveItem = null;
+
+        for (Treasure item : treasureHouse.getTreasures()) {
+            if (item.getTreasureCost() < leastExpensive) {
+                leastExpensive = item.getTreasureCost();
+                leastExpensiveItem = item;
+            }
+        }
+        return leastExpensiveItem;
+    }
+
+
+    public void findMostExpensiveTreasure(TreasureHouse treasureHouse) {
+        double mostExpensive = 0;
+        Treasure mostExpensiveItem = null;
+
+        for (Treasure item : treasureHouse.getTreasures()) {
+            if (item.getTreasureCost() > mostExpensive) {
+                mostExpensive = item.getTreasureCost();
+                mostExpensiveItem = item;
+            }
+        }
+        if (mostExpensiveItem != null) {
+            System.out.println("The most expensive item is " + mostExpensiveItem.toString());
+        } else {
+            System.out.println("There is no such item");
+        }
+    }
+
+    public void readTreasures(TreasureHouse treasureHouse) {
+        treasureHouse.setTreasures(reader.readObjectsFromFile(FILE_NAME));
+        showTreasures(treasureHouse.getTreasures());
+    }
+
+    private void showTreasures(List<Treasure> treasures) {
+        for (Treasure item : treasures) {
+            System.out.println((treasures.indexOf(item) + 1) + ". " + item.toString());
```

Comment:
If you need index please use another for loop. Call indexOf for every element is very expensive operation. What is complexity of this method?

Reply:
It's O(n2) if I'm not mistaken

Reply:
Yeah, which is really bad complexity for method which just prints elements.